provider: finalize EnvoyProxy referenced by a managed GatewayClass

cnvergence opened the pull request:
**What type of PR is this?**
<!--
Your PR title should be descriptive, and generally start with type that contains a subsystem name with `()` if necessary 
and summary followed by a colon. format `chore/docs/feat/fix/refactor/style/test: summary`.

Examples:
* "docs: fix grammar error"
* "feat(translator): add new feature"
* "fix: fix xx bug"
* "chore: change ci & build tools etc"
-->
Add support for envoy proxy finalizers  managed by GatewayClass

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Fixes:
- https://github.com/envoyproxy/gateway/issues/865

arkodg commented on internal/provider/kubernetes/controller_test.go:
```diff
@@ -89,6 +88,115 @@ func TestAddGatewayClassFinalizer(t *testing.T) {
+			expect: []string{"fooFinalizer"},
+		},
+		{
+			name: "envoyproxy with existing gatewayclass finalizer",
```
isn't this case same as L152 ?

cnvergence replied:
right, I will fix it

arkodg commented:
sorry for the delay in review @cnvergence, added some comments, can you please rebase as well, thanks !

cnvergence commented on internal/provider/kubernetes/controller.go:
```diff
@@ -249,7 +249,7 @@ func (r *gatewayAPIReconciler) Reconcile(ctx context.Context, request reconcile.
 
 	// Process the parametersRef of the accepted GatewayClass.
-	if acceptedGC.Spec.ParametersRef != nil && acceptedGC.DeletionTimestamp == nil {
+	if acceptedGC.Spec.ParametersRef != nil {
```
I have added the check inside of processParamsRef func to remove finalizer if gatewayClass will be terminated.
https://github.com/envoyproxy/gateway/blob/5797f6731a4846438a1372dfd78ca397ed36f36c/internal/provider/kubernetes/controller.go#L1344

arkodg replied:
the issue is at the end of this `if` where it performs a `r.resources.GatewayAPIResources.Store(acceptedGC.Name, resourceTree)`
we dont want to insert a deleted GC into the resource tree

cnvergence replied:
This `if` is used only for the `processParamsRef` function or am I missing something there?
```go
if acceptedGC.Spec.ParametersRef != nil {	
   if err := r.processParamsRef(ctx, acceptedGC, resourceTree); err != nil {
       msg := fmt.Sprintf("%s: %v", status.MsgGatewayClassInvalidParams, err)
       if err := r.gatewayClassUpdater(ctx, acceptedGC, false, string(gwapiv1b1.GatewayClassReasonInvalidParameters), msg); err != nil {
          r.log.Error(err, "unable to update GatewayClass status")
       }
       r.log.Error(err, "failed to process parametersRef for gatewayclass", "name", acceptedGC.Name)
       return reconcile.Result{}, err
   }
}
```
I moved the `acceptedGC.DeletionTimestamp` check inside this function, so if gateway class is being deleted `processParamsRef` will remove the finalizer from EnvoyProxy and return nil (and not add anything to the resourceTree).

arkodg replied:
yup you're right, this flow is fine, the `acceptedGC` at this point is not deleted since it has been checked earlier when computing `acceptedGC`, thanks for the correction

arkodg reviewed (approved):
LGTM thanks !

cnvergence commented:
@arkodg Yes, we should, but currently if GatewayClass stops referencing a EnvoyProxy, the EnvoyProxy resource will remain on the cluster.
I have opened up a follow-up issue for this during the work on this.
- https://github.com/envoyproxy/gateway/issues/1575

arkodg commented:
@cnvergence if we merge this PR, it feels like a regression for this case
* user creates EnvoyProxy E-A
* user attaches E-A to GatewayClass G-C, which in turn adds a finalizer on E-A
* user clears E-A from G-C
* since the finalizer is still on E-A, it cannot be deleted by user, which will have a negative impact on the user's experience

cnvergence commented:
/retest

cnvergence commented:
@arkodg I have changed `processParamsRef` function to support this case

cnvergence commented:
@arkodg understood, I have added this scenario to the integration test [here](https://github.com/envoyproxy/gateway/pull/1534/files#diff-5f29f313b6f74b9836b294d6fe03a909e7e023a42655873eb1a13b316e624e7dR212)
I have modified `processParamsRef` function so that whenever it won't match GatewayClass with EnvoyProxy https://github.com/envoyproxy/gateway/blob/e78e2dbb19cf140650ea38c5d668ba54c794177f/internal/provider/kubernetes/controller.go#L1464
and EnvoyProxy will have finalizers, they will be removed
https://github.com/envoyproxy/gateway/blob/e78e2dbb19cf140650ea38c5d668ba54c794177f/internal/provider/kubernetes/controller.go#L1487

cnvergence commented:
@arkodg understood, I have added this scenario to the kubernetes_test.go.
I have modified `processParamsRef` function so that whenever it won't match EnvoyProxy with GatewayClass and that EnvoyProxy will have finalizers, they will be removed.
https://github.com/envoyproxy/gateway/blob/e78e2dbb19cf140650ea38c5d668ba54c794177f/internal/provider/kubernetes/controller.go#L1464

arkodg commented on internal/provider/kubernetes/controller.go:
```diff
@@ -1432,30 +1458,40 @@ func (r *gatewayAPIReconciler) processParamsRef(ctx context.Context, gc *gwapiv1
 		}
-	}
 
-	if !found {
```
why have these been removed ?
these help build status messages that surface in the GatewayClass

cnvergence replied:
Right, I thought this change was needed and wanted to discuss it.
I have added it again and tested it out, it seems I was wrong earlier.
Thanks for catching this!

arkodg commented on internal/provider/kubernetes/controller.go:
```diff
@@ -1432,30 +1458,40 @@ func (r *gatewayAPIReconciler) processParamsRef(ctx context.Context, gc *gwapiv1
-		return fmt.Errorf("failed to find envoyproxy referenced by gatewayclass: %s", gc.Name)
-	}
-
-	if !valid {
```
same for this

arkodg commented on internal/provider/kubernetes/controller.go:
```diff
@@ -298,15 +298,13 @@ func (r *gatewayAPIReconciler) Reconcile(ctx context.Context, _ reconcile.Reques
-			}
-			r.log.Error(err, "failed to process parametersRef for gatewayclass", "name", acceptedGC.Name)
-			return reconcile.Result{}, err
+	if err := r.processParamsRef(ctx, acceptedGC, resourceTree); err != nil {
```
`if acceptedGC.Spec.ParametersRef != nil` is removed from here, can we add it back ?

arkodg reviewed (approved):
LGTM thanks @cnvergence !
